File: backend/src/main/kotlin/top/bearingwall/asya/service/UserService.kt
package top.bearingwall.asya.service

import jakarta.persistence.criteria.JoinType
import io.jsonwebtoken.JwtException
import org.slf4j.LoggerFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.domain.Specification
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import top.bearingwall.asya.audit.Auditable
import top.bearingwall.asya.dto.UserRegistrationRequest
import top.bearingwall.asya.dto.UserResponse
import top.bearingwall.asya.dto.UserInfoResponse
import top.bearingwall.asya.dto.UserUpdateRequest
import top.bearingwall.asya.dto.BatchRegisterRequest
import top.bearingwall.asya.dto.BatchRegisterResponse
import top.bearingwall.asya.dto.TokenRefreshResponse
import top.bearingwall.asya.model.User
import top.bearingwall.asya.model.UserRole
import top.bearingwall.asya.model.AuditActionType
import top.bearingwall.asya.repository.UserRepository
import top.bearingwall.asya.repository.ConferenceRepository
import top.bearingwall.asya.util.JwtUtil
import top.bearingwall.asya.util.TokenType
import java.util.UUID

@Service
class UserService(
    private val userRepository: UserRepository,
    private val systemConfigService: SystemConfigService,
    private val conferenceRepository: ConferenceRepository
) {

    private val log = LoggerFactory.getLogger(UserService::class.java)
    private val passwordEncoder = BCryptPasswordEncoder()

    @Transactional(readOnly = true)
    fun hasSysAdmin(): Boolean = userRepository.existsByRole(UserRole.SYS_ADMIN)

    @Transactional(readOnly = true)
    fun isRegistrationAvailable(): Boolean = !hasSysAdmin() || systemConfigService.isRegistrationAllowed()

    @Transactional
    @Auditable(type = AuditActionType.USER_REGISTER, content = "用户注册")
    fun registerUser(request: UserRegistrationRequest): AuthenticatedUserResponse {
        log.info("Registering user, name={}, role={}", request.name, request.role)

        if (!isRegistrationAvailable()) {
            throw IllegalStateException("系统当前禁止新用户注册")
        }

        val existing = userRepository.findByName(request.name)
        require(existing == null) { "User already exists: ${request.name}" }

        if (request.role == UserRole.SYS_ADMIN && hasSysAdmin()) {
            throw IllegalArgumentException("已经存在系统管理员，禁止重复注册")
        }

        val hashedPassword: String = requireNotNull(passwordEncoder.encode(request.password)) {
            "BCryptPasswordEncoder returned null hash"
        }

        val user = User(
            name = request.name,
            displayName = request.displayName,
            password = hashedPassword,
            role = request.role
        )

        val savedUser = userRepository.save(user)

        log.info("User registered successfully, uuid={}, name={}", savedUser.uuid, savedUser.name)

        return issueAuthenticatedUserResponse(savedUser)
    }

    @Auditable(type = AuditActionType.USER_LOGIN, content = "用户登录")
    fun loginUser(request: UserRegistrationRequest): AuthenticatedUserResponse {
        log.info("Logging in user, name={}", request.name)

        val user = userRepository.findByName(request.name)
            ?: throw IllegalStateException("系统中还没有该用户，请先完成注册")

        if (!passwordEncoder.matches(request.password, user.password)) {
            throw IllegalArgumentException("用户密码不正确")
        }

        log.info("User logged in successfully, uuid={}, name={}", user.uuid, user.name)

        return issueAuthenticatedUserResponse(user)
    }

    @Transactional(readOnly = true)
    fun refreshAccessToken(refreshToken: String): TokenRefreshResult {
        val user = getUserFromToken(refreshToken, TokenType.REFRESH)
        val tokens = issueTokens(user)
        return TokenRefreshResult(
            response = TokenRefreshResponse(token = tokens.accessToken),
            refreshToken = tokens.refreshToken
        )
    }

    // 获取所有用户（仅 SYS_ADMIN）
    @Transactional(readOnly = true)
    fun getAllUsers(): List<UserInfoResponse> {
        return userRepository.findAll()
            .map(::toUserInfoResponse)
    }

    @Transactional(readOnly = true)
    fun getUsers(
        pageable: Pageable,
        name: String?,
        displayName: String?,
        conferenceUuid: UUID?,
        role: UserRole?
    ): Page<UserInfoResponse> {
        val specification = Specification<User> { root, _, cb ->
            val predicates = mutableListOf<jakarta.persistence.criteria.Predicate>()

            name?.trim()?.takeIf { it.isNotEmpty() }?.let { keyword ->
                predicates += cb.like(cb.lower(root.get("name")), "%${keyword.lowercase()}%")
            }

            displayName?.trim()?.takeIf { it.isNotEmpty() }?.let { keyword ->
                predicates += cb.like(cb.lower(root.get("displayName")), "%${keyword.lowercase()}%")
            }

            conferenceUuid?.let {
                val conference = root.join<User, Any>("conference", JoinType.LEFT)
                predicates += cb.equal(conference.get<UUID>("uuid"), it)
            }

            role?.let {
                predicates += cb.equal(root.get<UserRole>("role"), it)
            }

            cb.and(*predicates.toTypedArray())
        }

        return userRepository.findAll(specification, pageable)
            .map(::toUserInfoResponse)
    }

    @Transactional
    @Auditable(type = AuditActionType.USER_DELETE, content = "删除用户")
    fun deleteUser(uuid: UUID) {
        if (!userRepository.existsById(uuid)) {
            throw IllegalArgumentException("User not found: $uuid")
        }
        // Potential: Handle associations (messages sent, messages received) if not handled by DB constraints/Cascades
        // For now, relying on JPA/DB configuration.
        userRepository.deleteById(uuid)
    }

    // 通过 token 获取当前登录用户信息
    @Transactional(readOnly = true)
    fun getCurrentUserInfo(token: String): UserInfoResponse {
        val user = getUserFromToken(token)
        return toUserInfoResponse(user)
    }

    @Transactional
    @Auditable(type = AuditActionType.USER_UPDATE, content = "更新用户信息")
    fun updateUser(targetUuid: UUID, token: String, request: UserUpdateRequest): UserInfoResponse {
        val requester = getUserFromToken(token)
        val target = userRepository.findById(targetUuid).orElseThrow {
            IllegalStateException("Target user not found")
        }
        // 权限校验：SYS_ADMIN 可以修改任何人；否则只能改自己
        val canEditAny = requester.role == UserRole.SYS_ADMIN
        val isSelf = requester.uuid == target.uuid
        require(canEditAny || isSelf) { "Permission denied" }

        // 应用更新字段
        request.name?.let { newName ->
            // 如果改名，确保唯一
            val existed = userRepository.findByName(newName)
            require(existed == null || existed.uuid == target.uuid) { "User name already exists" }
            target.name = newName
        }
        request.displayName?.let { newDisplayName ->
            target.displayName = newDisplayName
        }
        request.password?.let { newPassword ->
            val hashed: String = requireNotNull(passwordEncoder.encode(newPassword)) {
                "BCryptPasswordEncoder returned null hash"
            }
            target.password = hashed
            target.authVersion += 1
        }
        request.role?.let { newRole ->
            // 非管理员不能改角色除非是改自己的且不是提升特权，这里简单限制：只有管理员能改角色
            require(canEditAny) { "Only admin can change role" }
            target.role = newRole
        }

        val saved = userRepository.save(target)
        return toUserInfoResponse(saved)
    }

    @Transactional(readOnly = true)
    fun getUserFromToken(token: String, expectedType: TokenType = TokenType.ACCESS): User {
        val parsed = JwtUtil.parseToken(token)
        JwtUtil.requireTokenType(parsed, expectedType)
        val tokenVersion = JwtUtil.getAuthVersion(parsed)
        val userId = UUID.fromString(parsed.subject)
        val user = userRepository.findById(userId).orElseThrow {
            IllegalStateException("User not found by token subject")
        }
        if (user.authVersion != tokenVersion) {
            throw JwtException("Token已失效，请重新登录")
        }
        return user
    }

    @Transactional
    @Auditable(type = AuditActionType.USER_PASSWORD_RESET, content = "重置用户密码")
    fun resetPassword(uuid: UUID, newPassword: String) {
        val user = userRepository.findById(uuid).orElseThrow {
            IllegalArgumentException("User not found: $uuid")
        }
        val hashed: String = requireNotNull(passwordEncoder.encode(newPassword)) {
            "BCryptPasswordEncoder returned null hash"
        }
        user.password = hashed
        user.authVersion += 1
        userRepository.save(user)
    }

    @Transactional
    @Auditable(type = AuditActionType.USER_BATCH_REGISTER, content = "批量注册用户")
    fun batchRegister(requesterUuid: UUID, request: BatchRegisterRequest): BatchRegisterResponse {
        val requester = userRepository.findById(requesterUuid).orElseThrow { IllegalArgumentException("Requester not found") }
        if (requester.role != UserRole.SYS_ADMIN) {
            throw SecurityException("Only SYS_ADMIN can perform batch registration")
        }

        val conferenceUuid = UUID.fromString(request.conferenceId)
        val conference = conferenceRepository.findById(conferenceUuid).orElseThrow {
            IllegalArgumentException("Conference not found: ${request.conferenceId}")
        }

        val createdUsers = mutableListOf<UserInfoResponse>()

        for (item in request.users) {
            if (userRepository.findByName(item.name) != null) {
                throw IllegalArgumentException("User already exists: ${item.name}")
            }

            val hashedPassword: String = requireNotNull(passwordEncoder.encode(item.password)) {
                "BCryptPasswordEncoder returned null hash"
            }
            val user = User(
                name = item.name,
                displayName = item.displayName,
                password = hashedPassword,
                role = UserRole.DELEGATE,
                conference = conference
            )
            val saved = userRepository.save(user)
            createdUsers.add(toUserInfoResponse(saved))
        }

        return BatchRegisterResponse(createdUsers.size, createdUsers)
    }

    private fun toUserInfoResponse(user: User): UserInfoResponse {
        return UserInfoResponse(
            uuid = user.uuid?.toString() ?: "",
            name = user.name,
            displayName = user.displayName,
            role = user.role,
            conferenceUuid = user.conference?.uuid?.toString(),
            conferenceName = user.conference?.name
        )
    }

    private fun issueAuthenticatedUserResponse(user: User): AuthenticatedUserResponse {
        val tokens = issueTokens(user)
        return AuthenticatedUserResponse(
            response = UserResponse(
                uuid = user.uuid?.toString() ?: throw IllegalStateException("User id missing"),
                name = user.name,
                displayName = user.displayName,
                role = user.role,
                token = tokens.accessToken
            ),
            refreshToken = tokens.refreshToken
        )
    }

    private fun issueTokens(user: User): AuthTokenPair {
        val userId = user.uuid?.toString() ?: throw IllegalStateException("User id missing")
        val accessToken = JwtUtil.generateAccessToken(
            subject = userId,
            claims = mapOf("name" to user.name, "role" to user.role.name),
            authVersion = user.authVersion
        )
        val refreshToken = JwtUtil.generateRefreshToken(
            subject = userId,
            authVersion = user.authVersion
        )
        return AuthTokenPair(
            accessToken = accessToken,
            refreshToken = refreshToken
        )
    }
}

data class AuthenticatedUserResponse(
    val response: UserResponse,
    val refreshToken: String
)

data class TokenRefreshResult(
    val response: TokenRefreshResponse,
    val refreshToken: String
)

private data class AuthTokenPair(
    val accessToken: String,
    val refreshToken: String
)

File: backend-dotnet/src/AsyaMun.Api/Services/UserService.cs
using AsyaMun.Api.Auth;
using AsyaMun.Api.Data;
using AsyaMun.Api.Dtos;
using AsyaMun.Api.Errors;
using AsyaMun.Api.Models;
using Microsoft.EntityFrameworkCore;

namespace AsyaMun.Api.Services;

public record AuthenticatedUserResponse(UserResponse Response, string RefreshToken);

public record TokenRefreshResult(TokenRefreshResponse Response, string RefreshToken);

public class UserService
{
    private readonly AppDbContext _db;
    private readonly SystemConfigService _systemConfigService;

    public UserService(AppDbContext db, SystemConfigService systemConfigService)
    {
        _db = db;
        _systemConfigService = systemConfigService;
    }

    public async Task<bool> HasSysAdminAsync()
    {
        return await _db.Users.AsNoTracking().AnyAsync(u => u.Role == UserRole.SYS_ADMIN);
    }

    public async Task<bool> IsRegistrationAvailableAsync()
    {
        return !await HasSysAdminAsync() || await _systemConfigService.IsRegistrationAllowedAsync();
    }

    public async Task<AuthenticatedUserResponse> RegisterUserAsync(UserRegistrationRequest request)
    {
        if (!await IsRegistrationAvailableAsync())
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "系统当前禁止新用户注册");
        }

        var existing = await _db.Users.AsNoTracking().FirstOrDefaultAsync(u => u.Name == request.Name);
        if (existing != null)
        {
            throw new AsyaBusinessException(BizCode.USER_EXISTS, $"用户已存在: {request.Name}");
        }

        if (request.Role == UserRole.SYS_ADMIN && await HasSysAdminAsync())
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "已经存在系统管理员，禁止重复注册");
        }

        var user = new User
        {
            Name = request.Name,
            DisplayName = request.DisplayName,
            Password = BCrypt.Net.BCrypt.HashPassword(request.Password),
            Role = request.Role,
            AuthVersion = 0
        };

        _db.Users.Add(user);
        await _db.SaveChangesAsync();

        return IssueAuthenticatedUserResponse(user);
    }

    public async Task<AuthenticatedUserResponse> LoginUserAsync(UserRegistrationRequest request)
    {
        var user = await _db.Users
            .Include(u => u.Conference)
            .AsNoTracking()
            .FirstOrDefaultAsync(u => u.Name == request.Name);

        if (user == null)
        {
            throw new AsyaBusinessException(BizCode.USER_NOT_FOUND, "系统中还没有该用户，请先完成注册");
        }

        if (!BCrypt.Net.BCrypt.Verify(request.Password, user.Password))
        {
            throw new AsyaBusinessException(BizCode.PASSWORD_ERROR, "用户密码不正确");
        }

        return IssueAuthenticatedUserResponse(user);
    }

    public async Task<TokenRefreshResult> RefreshAccessTokenAsync(string refreshToken)
    {
        var user = await GetUserFromTokenAsync(refreshToken, TokenType.Refresh);
        var tokens = IssueTokens(user);

        return new TokenRefreshResult(
            new TokenRefreshResponse(tokens.AccessToken),
            tokens.RefreshToken);
    }

    public async Task<Page<UserInfoResponse>> GetUsersAsync(
        PageInput pageInput,
        string? name,
        string? displayName,
        Guid? conferenceUuid,
        UserRole? role,
        CancellationToken ct = default)
    {
        IQueryable<User> query = _db.Users.AsNoTracking().Include(u => u.Conference);

        if (!string.IsNullOrWhiteSpace(name))
        {
            var keyword = name.Trim().ToLower();
            query = query.Where(u => u.Name.ToLower().Contains(keyword));
        }

        if (!string.IsNullOrWhiteSpace(displayName))
        {
            var keyword = displayName.Trim().ToLower();
            query = query.Where(u => u.DisplayName != null && u.DisplayName.ToLower().Contains(keyword));
        }

        if (conferenceUuid.HasValue)
        {
            query = query.Where(u => u.ConferenceId == conferenceUuid.Value);
        }

        if (role.HasValue)
        {
            query = query.Where(u => u.Role == role.Value);
        }

        var effectiveInput = pageInput.Sort.Count == 0
            ? new PageInput { Page = pageInput.Page, Size = pageInput.Size, Sort = new List<SortSpec> { new("Name", false) } }
            : pageInput;

        var page = await query.ToPageAsync(effectiveInput, ct);

        return Page<UserInfoResponse>.Of(
            page.Content.Select(UserInfoResponse.From).ToList(),
            page.Number,
            page.Size,
            page.TotalElements);
    }

    public async Task DeleteUserAsync(Guid uuid)
    {
        var exists = await _db.Users.AsNoTracking().AnyAsync(u => u.Uuid == uuid);
        if (!exists)
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"用户不存在: {uuid}");
        }

        var user = await _db.Users.Include(u => u.MessageReceivers).FirstOrDefaultAsync(u => u.Uuid == uuid);
        if (user != null)
        {
            _db.MessageReceivers.RemoveRange(user.MessageReceivers);
            _db.Users.Remove(user);
            await _db.SaveChangesAsync();
        }
    }

    public UserInfoResponse GetCurrentUserInfo(User user)
    {
        return UserInfoResponse.From(user);
    }

    public async Task<UserInfoResponse> UpdateUserAsync(Guid targetUuid, User requester, UserUpdateRequest request)
    {
        var target = await _db.Users.Include(u => u.Conference)
            .FirstOrDefaultAsync(u => u.Uuid == targetUuid)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, "目标用户不存在");

        var canEditAny = requester.Role == UserRole.SYS_ADMIN;
        var isSelf = requester.Uuid == target.Uuid;
        if (!canEditAny && !isSelf)
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Permission denied", StatusCodes.Status403Forbidden);
        }

        if (request.Name != null)
        {
            var existed = await _db.Users.AsNoTracking().FirstOrDefaultAsync(u => u.Name == request.Name);
            if (existed != null && existed.Uuid != target.Uuid)
            {
                throw new AsyaBusinessException(BizCode.PARAM_ERROR, "用户名已存在");
            }

            target.Name = request.Name;
        }

        if (request.DisplayName != null)
        {
            target.DisplayName = request.DisplayName;
        }

        if (request.Password != null)
        {
            target.Password = BCrypt.Net.BCrypt.HashPassword(request.Password);
            target.AuthVersion += 1;
        }

        if (request.Role.HasValue)
        {
            if (!canEditAny)
            {
                throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Only admin can change role", StatusCodes.Status403Forbidden);
            }

            target.Role = request.Role.Value;
        }

        await _db.SaveChangesAsync();

        return UserInfoResponse.From(target);
    }

    public async Task ResetPasswordAsync(Guid uuid, string newPassword, Guid? requesterUuid)
    {
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Uuid == uuid)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"用户不存在: {uuid}");

        if (requesterUuid.HasValue)
        {
            var requester = await _db.Users.Include(u => u.Conference)
                .AsNoTracking()
                .FirstOrDefaultAsync(u => u.Uuid == requesterUuid.Value)
                ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"操作者不存在: {requesterUuid}");

            if (requester.Role == UserRole.DH)
            {
                if (requester.ConferenceId == null || requester.ConferenceId != user.ConferenceId)
                {
                    throw new ForbiddenException("DH 只能重置本会议内用户的密码");
                }
            }
        }

        user.Password = BCrypt.Net.BCrypt.HashPassword(newPassword);
        user.AuthVersion += 1;
        await _db.SaveChangesAsync();
    }

    public async Task<BatchRegisterResponse> BatchRegisterAsync(User requester, BatchRegisterRequest request)
    {
        if (requester.Role is not (UserRole.SYS_ADMIN or UserRole.DH))
        {
            throw new ForbiddenException("Only SYS_ADMIN or DH can perform batch registration");
        }

        if (!Guid.TryParse(request.ConferenceId, out var conferenceUuid))
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"会议ID无效: {request.ConferenceId}");
        }

        var conference = await _db.Conferences.FirstOrDefaultAsync(c => c.Uuid == conferenceUuid)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"会议不存在: {request.ConferenceId}");

        if (requester.Role == UserRole.DH)
        {
            if (requester.ConferenceId == null || requester.ConferenceId != conferenceUuid)
            {
                throw new ForbiddenException("DH can only batch register delegates to their own conference");
            }
        }

        await using var transaction = await _db.Database.BeginTransactionAsync();

        var createdUsers = new List<UserInfoResponse>();
        foreach (var item in request.Users)
        {
            var existing = await _db.Users.AsNoTracking().FirstOrDefaultAsync(u => u.Name == item.Name);
            if (existing != null)
            {
                throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"用户已存在: {item.Name}");
            }

            var user = new User
            {
                Name = item.Name,
                DisplayName = item.DisplayName,
                Password = BCrypt.Net.BCrypt.HashPassword(item.Password),
                Role = UserRole.DELEGATE,
                ConferenceId = conferenceUuid
            };

            _db.Users.Add(user);
            await _db.SaveChangesAsync();

            createdUsers.Add(UserInfoResponse.From(user));
        }

        await transaction.CommitAsync();

        return new BatchRegisterResponse(createdUsers.Count, createdUsers);
    }

    public async Task<BatchRegisterResponse> BatchRegisterFullAsync(User requester, BatchRegisterFullRequest request)
    {
        if (requester.Role is not (UserRole.SYS_ADMIN or UserRole.DH))
        {
            throw new ForbiddenException("Only SYS_ADMIN or DH can perform batch registration");
        }

        if (!Guid.TryParse(request.ConferenceId, out var conferenceUuid))
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"会议ID无效: {request.ConferenceId}");
        }

        var conference = await _db.Conferences.FirstOrDefaultAsync(c => c.Uuid == conferenceUuid)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"会议不存在: {request.ConferenceId}");

        if (requester.Role == UserRole.DH)
        {
            if (requester.ConferenceId == null || requester.ConferenceId != conferenceUuid)
            {
                throw new ForbiddenException("DH can only batch register users to their own conference");
            }
        }

        if (request.Users.Count == 0)
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "用户列表不能为空");
        }

        if (request.Users.Any(u => u.Role is not (UserRole.DELEGATE or UserRole.DM)))
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "批量注册仅支持 DELEGATE 或 DM 角色");
        }

        await using var transaction = await _db.Database.BeginTransactionAsync();

        var createdUsers = new List<UserInfoResponse>();
        var groupCache = new Dictionary<string, UserGroup>();
        var namesInBatch = new HashSet<string>();

        foreach (var item in request.Users)
        {
            var name = item.Name.Trim();
            var password = item.Password;

            if (name.Length == 0)
            {
                throw new AsyaBusinessException(BizCode.PARAM_ERROR, "用户昵称不能为空");
            }

            if (string.IsNullOrWhiteSpace(password))
            {
                throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"用户密码不能为空：{name}");
            }

            if (!namesInBatch.Add(name))
            {
                throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"用户名重复：{name}");
            }

            var existing = await _db.Users.AsNoTracking().FirstOrDefaultAsync(u => u.Name == name);
            if (existing != null)
            {
                throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"用户已存在: {name}");
            }

            var user = new User
            {
                Name = name,
                DisplayName = item.DisplayName,
                Password = BCrypt.Net.BCrypt.HashPassword(password),
                Role = item.Role,
                ConferenceId = conferenceUuid
            };

            var groupName = item.GroupName?.Trim();
            if (!string.IsNullOrEmpty(groupName))
            {
                if (!groupCache.TryGetValue(groupName, out var group))
                {
                    group = await _db.UserGroups.FirstOrDefaultAsync(g => g.GroupName == groupName);
                    if (group == null)
                    {
                        group = new UserGroup { GroupName = groupName };
                        _db.UserGroups.Add(group);
                    }

                    groupCache[groupName] = group;
                }

                group.UserUus.Add(user);
            }

            _db.Users.Add(user);
            await _db.SaveChangesAsync();

            createdUsers.Add(UserInfoResponse.From(user));
        }

        await transaction.CommitAsync();

        return new BatchRegisterResponse(createdUsers.Count, createdUsers);
    }

    public async Task<User> GetUserFromTokenAsync(string token, TokenType expectedType)
    {
        var parsed = JwtUtil.ParseToken(token);
        JwtUtil.RequireTokenType(parsed, expectedType);
        var tokenVersion = JwtUtil.GetAuthVersion(parsed);

        if (!Guid.TryParse(parsed.Subject, out var userId))
        {
            throw new TokenInvalidException("Token用户ID无效");
        }

        var user = await _db.Users.Include(u => u.Conference).FirstOrDefaultAsync(u => u.Uuid == userId)
            ?? throw new TokenInvalidException("用户不存在");

        if (user.AuthVersion != tokenVersion)
        {
            throw new TokenInvalidException("Token已失效，请重新登录");
        }

        return user;
    }

    public UserInfoResponse ToUserInfoResponse(User user)
    {
        return UserInfoResponse.From(user);
    }

    private AuthenticatedUserResponse IssueAuthenticatedUserResponse(User user)
    {
        var tokens = IssueTokens(user);

        return new AuthenticatedUserResponse(
            new UserResponse(
                user.Uuid.ToString("D"),
                user.Name,
                user.DisplayName,
                user.Role,
                tokens.AccessToken),
            tokens.RefreshToken);
    }

    private AuthTokenPair IssueTokens(User user)
    {
        var accessToken = JwtUtil.GenerateAccessToken(
            user.Uuid.ToString("D"),
            user.AuthVersion,
            user.Name,
            user.Role);

        var refreshToken = JwtUtil.GenerateRefreshToken(user.Uuid.ToString("D"), user.AuthVersion);

        return new AuthTokenPair(accessToken, refreshToken);
    }

    private sealed record AuthTokenPair(string AccessToken, string RefreshToken);
}